### PR TITLE
Fix use-of-uinitialized-value

### LINKED ISSUE #70263: MemorySanitizer: use-of-uninitialized-value in updateSerializationHintsForPart
https://s3.amazonaws.com/clickhouse-test-reports/69780/6ba4f61df86c8edc864863ef52a0a237bc5805b8/stateless_tests__msan__[2_4].html

Small snippet (full in https://s3.amazonaws.com/clickhouse-test-reports/69780/6ba4f61df86c8edc864863ef52a0a237bc5805b8/stateless_tests__msan__[2_4]/stderr.log.zst):
```
==593==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x560071b9cf2b in boost::multi_index::detail::ordered_index_node_compressed_base<boost::multi_index::detail::null_augment_policy, std::__1::allocator<char>>::parent_ref::operator boost::multi_index::detail::ordered_index_node_impl<boost::multi_index::detail::null_augment_policy, std::__1::allocator<char>>*() const build_docker/./contrib/boost/boost/multi_index/detail/ord_index_node.hpp:176:31
    #1 0x560071b9cf2b in boost::multi_index::detail::ordered_index_impl<boost::multi_index::member<DB::ColumnDescription, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, &DB::ColumnDescription::name>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, boost::multi_index::detail::nth_layer<2, DB::ColumnDescription, boost::multi_index::indexed_by<boost::multi_index::sequenced<boost::multi_index::tag<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>>, boost::multi_index::ordered_unique<boost::multi_index::member<DB::ColumnDescription, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, &DB::ColumnDescription::name>, mpl_::na, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, std::__1::allocator<DB::ColumnDescription>>, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_unique_tag, boost::multi_index::detail::null_augment_policy>::root() const build_docker/./contrib/boost/boost/multi_index/detail/ord_index_impl.hpp:1133:40
    #2 0x560071b9cf2b in boost::multi_index::detail::bidir_node_iterator<boost::multi_index::detail::ordered_index_node<boost::multi_index::detail::null_augment_policy, boost::multi_index::detail::index_node_base<DB::ColumnDescription, std::__1::allocator<DB::ColumnDescription>>>> boost::multi_index::detail::ordered_index_impl<boost::multi_index::member<DB::ColumnDescription, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, &DB::ColumnDescription::name>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, boost::multi_index::detail::nth_layer<2, DB::ColumnDescription, boost::multi_index::indexed_by<boost::multi_index::sequenced<boost::multi_index::tag<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>>, boost::multi_index::ordered_unique<boost::multi_index::member<DB::ColumnDescription, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, &DB::ColumnDescription::name>, mpl_::na, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, std::__1::allocator<DB::ColumnDescription>>, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_unique_tag, boost::multi_index::detail::null_augment_policy>::find<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const build_docker/./contrib/boost/boost/multi_index/detail/ord_index_impl.hpp:611:45
    #3 0x56007b646c26 in DB::ColumnsDescription::tryGetColumn(DB::GetColumnsOptions const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const build_docker/./src/Storages/ColumnsDescription.cpp:700:32
    #4 0x56007b64a0fb in DB::ColumnsDescription::tryGetPhysical(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const build_docker/./src/Storages/ColumnsDescription.cpp:771:12
    #5 0x56007ceae801 in void DB::updateSerializationHintsForPart<std::__1::shared_ptr<DB::IMergeTreeDataPart>>(std::__1::shared_ptr<DB::IMergeTreeDataPart> const&, DB::ColumnsDescription const&, DB::SerializationInfoByName&, bool) build_docker/./src/Storages/MergeTree/MergeTreeData.cpp:8589:66
    #6 0x56007ceae801 in void DB::MergeTreeData::updateSerializationHints<std::__1::set<std::__1::shared_ptr<DB::IMergeTreeDataPart>, DB::MergeTreeData::LessDataPart, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart>>>, std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>>(std::__1::set<std::__1::shared_ptr<DB::IMergeTreeDataPart>, DB::MergeTreeData::LessDataPart, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart>>> const&, std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, DB::DataPartsLock const&) build_docker/./src/Storages/MergeTree/MergeTreeData.cpp:8622:9
    #7 0x56007ce162e7 in DB::MergeTreeData::Transaction::commit(DB::DataPartsLock*) build_docker/./src/Storages/MergeTree/MergeTreeData.cpp:6856:9
    #8 0x56007d9a76d1 in DB::MergePlainMergeTreeTask::finish() build_docker/./src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:146:17
    #9 0x56007d9a4853 in DB::MergePlainMergeTreeTask::executeStep() build_docker/./src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:71:13
    #10 0x56007d8f8842 in DB::executeHere(std::__1::shared_ptr<DB::MergePlainMergeTreeTask>) build_docker/./src/Storages/MergeTree/MergePlainMergeTreeTask.h:104:18
    #11 0x56007d8f8842 in DB::StorageMergeTree::merge(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, bool, std::__1::shared_ptr<DB::MergeTreeTransaction> const&, PreformattedMessage&, bool) build_docker/./src/Storages/StorageMergeTree.cpp:1195:5
    #12 0x56007d90e611 in DB::StorageMergeTree::optimize(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::IAST> const&, bool, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, bool, std::__1::shared_ptr<DB::Context const>) build_docker/./src/Storages/StorageMergeTree.cpp:1626:18
    #13 0x560078f25bd0 in DB::InterpreterOptimizeQuery::execute() build_docker/./src/Interpreters/InterpreterOptimizeQuery.cpp:84:12
    #14 0x560078e09de5 in DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/./src/Interpreters/executeQuery.cpp:1318:40
    #15 0x560078dfb0dc in DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) build_docker/./src/Interpreters/executeQuery.cpp:1485:26
    #16 0x56007e19a057 in DB::TCPHandler::runImpl() build_docker/./src/Server/TCPHandler.cpp:587:54
    #17 0x56007e1ec130 in DB::TCPHandler::run() build_docker/./src/Server/TCPHandler.cpp:2520:9
    #18 0x560083d904bf in Poco::Net::TCPServerConnection::start() build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43:3
    #19 0x560083d913ff in Poco::Net::TCPServerDispatcher::run() build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115:38
    #20 0x560083c72114 in Poco::PooledThread::run() build_docker/./base/poco/Foundation/src/ThreadPool.cpp:205:14
    #21 0x560083c6eaed in Poco::(anonymous namespace)::RunnableHolder::run() build_docker/./base/poco/Foundation/src/Thread.cpp:45:11
    #22 0x560083c6b9b0 in Poco::ThreadImpl::runnableEntry(void*) build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:335:27
    #23 0x7f8362997ac2  (/lib/x86_64-linux-gnu/libc.so.6+0x94ac2) (BuildId: 490fef8403240c91833978d494d39e537409b92e)
    #24 0x7f8362a2984f  (/lib/x86_64-linux-gnu/libc.so.6+0x12684f) (BuildId: 490fef8403240c91833978d494d39e537409b92e)

  Uninitialized value was created by a heap deallocation
    #0 0x56004b2570e9 in operator delete(void*, unsigned long) (/usr/bin/clickhouse+0x84370e9) (BuildId: ba4d5c490217edfc28d0781822fd6edc67a199f9)
    #1 0x56007132575e in std::__1::default_delete<DB::StorageInMemoryMetadata const>::operator()[abi:v15007](DB::StorageInMemoryMetadata const*) const build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48:5
    #2 0x56007132575e in std::__1::__shared_ptr_pointer<DB::StorageInMemoryMetadata const*, std::__1::default_delete<DB::StorageInMemoryMetadata const>, std::__1::allocator<DB::StorageInMemoryMetadata const>>::__on_zero_shared() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:263:5
    #3 0x560071325312 in std::__1::__shared_count::__release_shared[abi:v15007]() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:174:9
    #4 0x560071325312 in std::__1::__shared_weak_count::__release_shared[abi:v15007]() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:215:27
    #5 0x560071325312 in std::__1::shared_ptr<DB::StorageInMemoryMetadata const>::~shared_ptr[abi:v15007]() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:702:23
    #6 0x560071325312 in std::__1::shared_ptr<DB::StorageInMemoryMetadata const>::operator=[abi:v15007](std::__1::shared_ptr<DB::StorageInMemoryMetadata const>&&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:723:9
    #7 0x560071325312 in MultiVersion<DB::StorageInMemoryMetadata>::set(std::__1::unique_ptr<DB::StorageInMemoryMetadata const, std::__1::default_delete<DB::StorageInMemoryMetadata const>>&&) build_docker/./src/Common/MultiVersion.h:76:25
    #8 0x56007131ffb8 in DB::IStorage::setInMemoryMetadata(DB::StorageInMemoryMetadata const&) build_docker/./src/Storages/IStorage.h:218:18
    #9 0x56007cd5e815 in DB::MergeTreeData::changeSettings(std::__1::shared_ptr<DB::IAST> const&, std::__1::unique_lock<std::__1::timed_mutex>&) build_docker/./src/Storages/MergeTree/MergeTreeData.cpp:3747:9
    #10 0x56007d8d20f0 in DB::StorageMergeTree::alter(DB::AlterCommands const&, std::__1::shared_ptr<DB::Context const>, std::__1::unique_lock<std::__1::timed_mutex>&) build_docker/./src/Storages/StorageMergeTree.cpp:365:9
    #11 0x5600773decb3 in DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) build_docker/./src/Interpreters/InterpreterAlterQuery.cpp:214:16
    #12 0x5600773d6d2f in DB::InterpreterAlterQuery::execute() build_docker/./src/Interpreters/InterpreterAlterQuery.cpp:75:16
    #13 0x560078e09de5 in DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/./src/Interpreters/executeQuery.cpp:1318:40
    #14 0x560078dfb0dc in DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) build_docker/./src/Interpreters/executeQuery.cpp:1485:26
    #15 0x56007e19a057 in DB::TCPHandler::runImpl() build_docker/./src/Server/TCPHandler.cpp:587:54
    #16 0x56007e1ec130 in DB::TCPHandler::run() build_docker/./src/Server/TCPHandler.cpp:2520:9
    #17 0x560083d904bf in Poco::Net::TCPServerConnection::start() build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43:3
    #18 0x560083d913ff in Poco::Net::TCPServerDispatcher::run() build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115:38
    #19 0x560083c72114 in Poco::PooledThread::run() build_docker/./base/poco/Foundation/src/ThreadPool.cpp:205:14
    #20 0x560083c6eaed in Poco::(anonymous namespace)::RunnableHolder::run() build_docker/./base/poco/Foundation/src/Thread.cpp:45:11
    #21 0x560083c6b9b0 in Poco::ThreadImpl::runnableEntry(void*) build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:335:27
    #22 0x7f8362997ac2  (/lib/x86_64-linux-gnu/libc.so.6+0x94ac2) (BuildId: 490fef8403240c91833978d494d39e537409b92e)

```

Seems like it might have been introduced in https://github.com/ClickHouse/ClickHouse/pull/69828 cc @CurtizJ

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Closes #70263.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 90746cbb4aedfe768d8faf404bf4c1eed082d27c with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70273/90746cbb4aedfe768d8faf404bf4c1eed082d27c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/89d4ab65e6749864bf1aa7c587d6685a3e91269a/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -8605,7 +8605,9 @@ void MergeTreeData::resetSerializationHints(const DataPartsLock & /*lock*/)
         .choose_kind = true,
     };
 
-    const auto & storage_columns = getInMemoryMetadataPtr()->getColumns();
+    const auto meta_data = getInMemoryMetadataPtr();
```
Comment (Algunenano) [inline src/Storages/MergeTree/MergeTreeData.cpp:8608]: What about the other uses of `getInMemoryMetadataPtr()->getColumns()`? In fact the associated issue complains about `updateSerializationHints`

Comment (CurtizJ) [reply src/Storages/MergeTree/MergeTreeData.cpp:8608]: Yes, please update the same case a few lines below.

Comment (Algunenano) [reply src/Storages/MergeTree/MergeTreeData.cpp:8608]: Suspicious in the same file: `updateSerializationHints` and `addPartContributionToColumnAndSecondaryIndexSizes` (but this one is not new/recent)

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**